### ISSUE TITLE
How to load a metagraph via C++

### ISSUE THREAD

**Issue by Kika-Xumengwei** [tensorflow/tensorflow#15687]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug or a feature request.
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.2
- **Python version**: 2.7

### Describe the problem
I want to load a metagraph via C++ code, and later the checkpoint weights. To load the metagraph, I first generate a pb file from it

```
with tf.Session() as sess:
	new_saver = tf.train.import_meta_graph(root_dir + meta_graph)
	tf.train.write_graph(sess.graph_def, root_dir, export_pb, as_text=False)
```
Then I use selective registration to generate the ops and kernels needed

`bazel-bin/tensorflow/python/tools/print_selective_registration_header --graphs=xinmei/rnn_dict/model_test.pb > tensorflow/core/framework/ops_to_register.h`
Next, I compile my runnable using this registration. However, when I run the executable on my Android device, it says

```
Error creating graph: Invalid argument: No OpKernel was registered to support Op 'Const' with these attrs.  Registered devices: [CPU], Registered kernels:
  <no registered kernels>

	 [[Node: save/RestoreV2_8/shape_and_slices = Const[_output_shapes=[[1]], dtype=DT_STRING, value=Tensor<type: string shape: [1] values: >]()]]
```
How can I load the metagraph? The motivation of this is that I want to continue training the model on my Android device.

**Comment by bignamehyp** [comment]
metagraph contains some python specific objects that can't be loaded with c++ currently.
@mrry, anything to add or correct?

**Comment by mrry** [comment]
While there is no specific API for loading a `MetaGraphDef` in C++, you can either package it as a SavedModel and [load it in C++](https://www.tensorflow.org/programmers_guide/saved_model#loading_a_savedmodel_in_c) or extract the [`GraphDef`](https://github.com/tensorflow/tensorflow/blob/271917a124a5b86c9ff6444415d20402a5db4ac0/tensorflow/core/protobuf/meta_graph.proto#L68) from inside the `MetaGraphDef` and load that directly.

**Comment by xumengwei** [comment]
@bignamehyp @mrry Thanks for your replies. I actually tried the second way you suggested: just loading the pb file rather than the meta file, but it says

`[libprotobuf ERROR external/protobuf/src/google/protobuf/wire_format_lite.cc:621] String field 'tensorflow.MetaGraphDef.MetaInfoDef.tensorflow_version' contains invalid UTF-8 data when parsing a protocol buffer. Use the 'bytes' type if you intend to send raw bytes. Error reading graph definition from model_test.pb: Data loss: Can't parse model_test.pb as binary proto`

Why comes this error?

Btw, I assume it is feasible that I load the graph & weights for continuing training on a Android device?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by drpngx** [comment]
@xumengwei the common cause is either you have the wrong proto, or you're trying to load a binary as text.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by drpngx** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
